exp/runtime/internal/controllers/extensionconfig_controller_test.go

```diff
@@ -0,0 +1,304 @@
+
+	cat := catalog.New()
+	registry := runtimev1registry.New()
+	runtimeClient := runtimeclient.New(runtimeclient.Options{
+		Catalog:  cat,
+		Registry: registry,
+	})
+
+	g.Expect(runtimehooksv1.AddToCatalog(cat)).To(Succeed())
+
+	r := &Reconciler{
+		Client:        env.GetClient(),
+		APIReader:     env.GetAPIReader(),
+		RuntimeClient: runtimeClient,
+	}
+
+	server := fakeExtensionServer(discoveryHandler("first", "second", "third"))
+	extensionConfig := fakeExtensionConfigForURL(ns.Name, "ext1", server.URL)
+	defer server.Close()
+	// Create the ExtensionConfig.
+	g.Expect(env.CreateAndWait(ctx, extensionConfig)).To(Succeed())
+	defer func() {
+		g.Expect(env.CleanupAndWait(ctx, extensionConfig)).To(Succeed())
+	}()
+	t.Run("fail reconcile if registry has not been warmed up", func(t *testing.T) {
+		// Attempt to reconcile. This will be an error as the registry has not been warmed up at this point.
+		res, err := r.Reconcile(ctx, ctrl.Request{NamespacedName: util.ObjectKey(extensionConfig)})
+		g.Expect(err).To(BeNil())
+		// If the registry isn't warm the reconcile loop will return Requeue: True
+		g.Expect(res.Requeue).To(Equal(true))
+	})
+
+	t.Run("successful reconcile and discovery on ExtensionConfig create", func(t *testing.T) {
+		// Warm up the registry before trying reconciliation again.
+		warmup := &warmupRunnable{
+			Client:        env.GetClient(),
+			APIReader:     env.GetAPIReader(),
+			RuntimeClient: runtimeClient,
+		}
+		g.Expect(warmup.Start(ctx)).To(Succeed())
+
+		// Reconcile the extension and assert discovery has succeeded.
+		_, err = r.Reconcile(ctx, ctrl.Request{NamespacedName: util.ObjectKey(extensionConfig)})
+		g.Expect(err).To(BeNil())
+
+		config := &runtimev1.ExtensionConfig{}
+		g.Expect(env.GetAPIReader().Get(ctx, util.ObjectKey(extensionConfig), config)).To(Succeed())
+
+		// Expect three handlers for the extension and expect the name to be the handler name plus the extension name.
+		handlers := config.Status.Handlers
+		g.Expect(len(handlers)).To(Equal(3))
+		g.Expect(handlers[0].Name).To(Equal("first.ext1"))
+		g.Expect(handlers[1].Name).To(Equal("second.ext1"))
+		g.Expect(handlers[2].Name).To(Equal("third.ext1"))
+
+		conditions := config.GetConditions()
+		g.Expect(len(conditions)).To(Equal(1))
+		g.Expect(conditions[0].Status).To(Equal(corev1.ConditionTrue))
+		g.Expect(conditions[0].Type).To(Equal(runtimev1.RuntimeExtensionDiscoveredCondition))
+		_, err = registry.Get("first.ext1")
+		g.Expect(err).To(BeNil())
+		_, err = registry.Get("second.ext1")
+		g.Expect(err).To(BeNil())
+		_, err = registry.Get("third.ext1")
+		g.Expect(err).To(BeNil())
+	})
+
+	t.Run("Successful reconcile and discovery on Extension update", func(t *testing.T) {
+		// Start a new ExtensionServer where the second handler is removed.
+		updatedServer := fakeExtensionServer(discoveryHandler("first", "third"))
+		defer updatedServer.Close()
+		// Close the original server  it's no longer serving.
+		server.Close()
```

Comment:
Why do we need this while we have `defer server.Close()` at L69

Reply:
Just making sure the original server isn't returning any result when contacted for this test. It's not about cleanup, but instead making sure we get a connection error if the config is out of date.